lib/perl/Genome/Sys/ShellPipeline.pm

```diff
@@ -0,0 +1,190 @@
+
+class Genome::Sys::ShellPipeline {
+    is => "UR::Object",
+    attributes_have => [qw(is_input is_output is_optional)],
+    has_input => {
+        pre_commands => {
+            is => "ARRAY",
+            doc => "Arrayref of commands to execute before the main pipe commands",
+            default_value => [],
+        },
+
+        post_commands => {
+            is => "ARRAY",
+            doc => "Arrayref of commands to execute after the main pipe commands",
+            default_value => [],
+        },
+
+        pipe_commands => {
+            is => "ARRAY",
+            doc => "Arrayref of commands to be piped together",
+        },
+
+        redirects => {
+            is => "Text",
+            doc => "If specified, this text is appended to the concatenated pipe " .
+                "commands without a pipe inbetween (e.g., > /tmp/stdout.txt 2>> /tmp/log)",
+            is_optional => 1,
+        },
+
+        interpreter => {
+            is => "Text",
+            doc => "The interpreter to use to run the pipe script",
+            default_value => "/bin/bash",
+        },
+
+        script_path => {
+            is => "Text",
+            doc => "If specified, write the pipe script to this path. " .
+                "By default, it is written to a temporary location.",
+            is_optional => 1,
+        },
+    },
+
+    has_output => [
+        return_codes => {
+            is => "ARRAY",
+            doc => "The return code of each command in pipe_commands",
+        },
+    ]
+};
+
+sub _validate_params {
+    my $self = shift;
+
+    my $commands = $self->pipe_commands;
+    croak $self->error_message("pipe_commands argument must be a nonempty arrayref")
+        unless (ref $commands eq "ARRAY" && scalar @$commands > 0);
+
+    for my $x (qw(pre post)) {
+        my $var = sprintf "%s_commands", $x;
+        my $value = $self->$var;
+        croak $self->error_message("$var argument must be undef or an arrayref")
+            unless !defined $value || ref $value eq "ARRAY";
+    }
+
+}
+
+sub execute {
+    my $self = shift;
+
+    $self->_validate_params;
+
+    my $script_path = $self->script_path;
+    my $tmpdir = Genome::Sys->create_temp_directory;
+    if (!defined $script_path) {
+        $script_path = File::Spec->catfile($tmpdir, "pipe.sh");
+    }
+
+    my $pipestatus_path = File::Spec->catfile($tmpdir, "pipe.status");
+
+    write_file($script_path, $self->script_text($pipestatus_path));
+
+    $self->debug_message(
+        sprintf
+            "Executing script $script_path:\n\n" .
+            "-------- BEGIN SCRIPT --------\n%s" .
+            "--------- END SCRIPT ---------\n",
+            join("", read_file($script_path))
+        );
+
+
+    chmod oct(755), $script_path;
+
+    try {
+        Genome::Sys->shellcmd(cmd => $script_path);
+    }
+    catch {
+        croak $self->error_message($_);
+    }
+    finally {
+        $self->_check_pipestatus($pipestatus_path);
```

Comment:
Since `_check_pipestatus` wants to throw exceptions I'm not sure this is a safe use of `finally`.  From the docs:

> Furthermore exceptions in finally blocks are not trappable and are unable to influence the execution of your program. This is due to limitation of DESTROY-based scope guards, which finally is implemented on top of. This may change in a future version of Try::Tiny.

I think `_check_pipestatus` can just be moved inside the `try` block?